dadi/lib/providers/restapi.js

```diff
@@ -0,0 +1,240 @@
+    provider: this.provider,
+    config: this.config,
+    defaults: this.datasource.source.auth
+  })
+}
+
+/**
+ * load - loads data form the datasource
+ *
+ * @param  {string} requestUrl - url of the web request (not used)
+ * @param  {fn} done - callback on error or completion
+ * @return {void}
+ */
+RestApi.prototype.load = function load (requestUrl, done) {
+  log.info(
+    { module: 'restapi' },
+    'GET datasource "' + this.datasource.schema.datasource.key + '"'
+  )
+
+  // allow query string param to bypass cache
+  var query = url.parse(requestUrl, true).query
+  var noCache = query.cache && query.cache.toString().toLowerCase() === 'false'
+
+  try {
+    const endpoint = this.schema.datasource.source.endpoint
+    const query = this.query
+
+    this.cacheKey = [
+      endpoint,
+      encodeURIComponent(JSON.stringify(this.schema.datasource))
+    ].join('+')
+
+    this.dataCache = new DatasourceCache()
+
+    var cacheOptions = {
+      name: this.datasource.name,
+      caching: this.schema.datasource.caching,
+      cacheKey: this.cacheKey
+    }
+
+    this.dataCache.getFromCache(cacheOptions, cachedData => {
+      if (cachedData && !noCache) {
+        try {
+          cachedData = JSON.parse(cachedData.toString())
+          return done(null, cachedData)
+        } catch (err) {
+          console.error(
+            'Rest provider: cache data incomplete, making HTTP request: ' +
+              err +
+              '(' +
+              cacheOptions.cacheKey +
+              ')'
+          )
+        }
+      }
+
+      var self = this
+
+      this.purest
+        .select(endpoint)
+        .where(query)
+        .request(function (err, res, body) {
+          log.info(
+            { module: 'restapi' },
+            'GOT datasource "' +
+              self.datasource.schema.datasource.key +
+              '": ' +
+              decodeURIComponent(res ? res.request.href : '') +
+              ' (HTTP 200, ' +
+              help.formatBytes(Buffer.byteLength(body)) +
+              ')'
+          )
+
+          self.processOutput(res, err, body, noCache, done)
+        })
+    })
+  } catch (ex) {
+    done(ex, null)
+  }
+}
+
+/**
+ * processOutput
+ *
+ * @param  {response} res
+ * @param  {string} data
+ * @param  {fn} done
+ * @return {void}
+ */
+RestApi.prototype.processOutput = function processOutput (
+  res,
+  error,
+  data,
+  noCache,
+  done
+) {
+  if (res.statusCode === 200) {
+    data = this.processFields(data)
+
+    var cacheOptions = {
+      name: this.datasource.name,
+      caching: this.schema.datasource.caching,
+      cacheKey: this.cacheKey
+    }
+
+    if (!noCache) {
+      this.dataCache.cacheResponse(cacheOptions, JSON.stringify(data), () => {
+        //
+      })
+    }
+  }
+
+  if (res.statusCode === 404) {
+    data = {
+      results: [],
+      errors: [
+        {
+          code: 'WEB-0004',
+          title: 'Datasource Not Found',
+          details:
```

Comment:
Have you considered using template strings here? I think it would read better.

```js
{
  details: `Datasource "${this.datasource.name}" failed. ${res.statusMessage} (${res.statusCode}): ${res.request.href}`
}
```

Reply:
Agree this is cleaner, but it matches similar errors elsewhere. @jimlambie talked about refactoring how these errors are returned from providers  (#377) so I might leave this until then